guardian/pinboard
# Desktop notifications

_NOTE: this branch has been rebased on top of `persisting-seen-by` (to sort out merge conflicts proactively) so it's highly recommended to merge https://github.com/guardian/editorial-tools-pinboard/pull/57 before this PR (at which point the base branch of this PR should automatically change to `main`)._

---

**Co-authored-by: @twrichards** 

https://trello.com/c/KFTGgSoV/744-desktop-notifications

Users have asked that they have desktop notifications for when they're either not using any tool windows, or are busy doing something else with at least one tab open elsewhere.

## What does this change?
- Introduces a new `notifications-lambda`, which is triggered by a 'DynamoDB Stream' from the Items table (batched in a 10second window). The lambda first queries the User table to retrieve a list of users who have the 🆕 `webPushNotification` column populated (see below bullet point) and then loops through the Item records it received in the input event/payload and if the user in the item is 'mentioned' then sends a [push notification](https://developer.mozilla.org/en-US/docs/Web/API/Push_API) using the [`web-push`](https://www.npmjs.com/package/web-push) library.
_NOTE: in future, once we've done more thinking about what constitutes a relevant pinboard and implemented a 'following' feature, we can send more notifications than just mentions._

- Introduced a [Service Worker](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) on the pinboard domain (rather than one for each domain that pinboard is embedded into, currently composer and grid, but more in future 🤞 , as there's no way to de-duplicated notifications across domains - so the notifications must originate from the same exact domain, and thus must originate from a single service worker), the service worker also lives on beyond any of the tabs, so it can still receive web push notifications, so long as the browser process is running 🎉 .

- Introduced a new field `webPushNotification` on the User table (with associated AppSync/GraphQL queries & mutations) which holds the details of the the 'web push subscription' generated by the client, at the point they subscribe for Desktop Notifications. Due to some cross-origin & other protections in the browser this has to be done in a pop-out window to the pinbord domain, and works as follows...
![pinboard desktop notifications diagram (part 1)](https://user-images.githubusercontent.com/19289579/132382838-216f772d-534d-426c-bad9-adcf1caa98ef.png)
https://docs.google.com/drawings/d/14GGykKg_GsssXLO9vcL08z_b4IQ0QzPkQ-nVNDfnaDU/edit?usp=sharing

- Since we only send 'mention' notifications from the `notifications-lambda` we also want to display desktop notifications when they have composer, grid etc. tab(s) open and items are received using the existing AppSync/GraphQL subscriptions, to do this we need send these to service worker (so that it can de-duplicate those and so they originate from the pinboard domain and so we don't need notification permission on all the various tool domains) - in order to do this we need to `postMessage` to a hidden iFrame (which is displaying a page on the pinboard domain) which then does a further `postMessage` on to the service worker so it can show the notification. This complicated process is best explained via this diagram which shows the sequence...
![pinboard desktop notifications diagram (part 2)](https://user-images.githubusercontent.com/19289579/132404735-7f68ada7-26d5-4f8c-b34d-6c7d74a6f7db.png)
https://docs.google.com/drawings/d/17--IeZ5GCUeWkPuDnfV_3anyk9g0CwwbZoJ6Q4w_M8I/edit?usp=sharing

## How to test

1. Deploy this to CODE
2. (if you haven't already) click the `Subscribe to Desktop Notifications` button at the top of the list of available pinboards (and then again on a similar button in the pop-out tab)
![image](https://user-images.githubusercontent.com/19289579/132493050-e28d06bb-8e5c-4580-9e25-346f7083eccd.png)
3. and then with the help of another person, have them send some messages
    - both non-mentions (for which you will need at least one composer tab open)
    - mentions (which should still appear without any tools windows open)

## How can we measure success?
This fulfills a direct feature request from users, but is generally useful and arguably an expected part of a modern real-time messaging/sharing system.

## Images
The below shows to messages and their corresponding desktop notifications (one of which has a grid-crop attached, and we see the thumbnail in the notification).
![image](https://user-images.githubusercontent.com/19289579/132491707-81715ad6-b84a-4f57-9127-f5f91b8406d9.png)
